python/cudf/cudf/core/index.py

```diff
@@ -2796,59 +2802,71 @@ def interval_range(
     >>> cudf.interval_range(start=0,end=5)
     IntervalIndex([(0, 0], (1, 1], (2, 2], (3, 3], (4, 4], (5, 5]],
     ...closed='right',dtype='interval')
-
     >>> cudf.interval_range(start=0,end=10, freq=2,closed='left')
     IntervalIndex([[0, 2), [2, 4), [4, 6), [6, 8), [8, 10)],
     ...closed='left',dtype='interval')
-
     >>> cudf.interval_range(start=0,end=10, periods=3,closed='left')
     ...IntervalIndex([[0.0, 3.3333333333333335),
             [3.3333333333333335, 6.666666666666667),
             [6.666666666666667, 10.0)],
             closed='left',
             dtype='interval')
-
     """
     if freq and periods and start and end:
         raise ValueError(
             "Of the four parameters: start, end, periods, and "
             "freq, exactly three must be specified"
         )
+    if isinstance(start or end, dt.datetime) or isinstance(
+        start or end, dt.timedelta
+    ):
+        raise NotImplementedError(
+            "Datetime-like intervals are not yet supported"
+        )
+    if start and isinstance(start, str):
+        raise ValueError(f"start must be numeric, got {start}")
+    if end and isinstance(end, str):
+        raise ValueError(f"end must be numeric, got {start}")
     elif periods and not freq:
-        assert end is not None and start is not None
-        end = end + 1
-        periods_array = cupy.asarray(arange(start, end))
-        mn, mx = (periods_array.min(), periods_array.max())
-        bin_edges = cupy.linspace(mn, mx, periods + 1, endpoint=True)
-        # cupy.linspace turns all arrays into a float array
-        # this can cause the dtype to be a float instead of an int
-        # the below adjusts for this
-        if cupy.all(cupy.mod(bin_edges, 1) == 0):
-            bin_edges = bin_edges.astype(int)
-        left_col = bin_edges[:-1]
-        right_col = bin_edges[1:]
+        # if statement for mypy to pass
+        if end is not None and start is not None:
+            freq_step = ((end) - start) / periods
+            freq_step = DeviceScalar(freq_step, dtype="float64")
+            start = DeviceScalar(start, dtype="float64")
+            bin_edges = sequence(size=periods + 1, init=start, step=freq_step)
+            if cupy.all(cupy.mod(bin_edges, 1) == 0):
+                bin_edges = bin_edges.astype(int)
```

Comment:
Clarifying that I was using this code to figure out if all the numbers in `bin_edges` were integers.
 Using `sequence` in a way that would work gives us a result column ,`bin_edges`, with a dtype of `float`. 

For example we get `[5.0, 2.0, 10.0]` instead of `[5, 2, 10]` which then gives our overall index a dtype of `float` instead of `int`.

The final dtype is determined by the dtype of the `bin_edges`, so typcasting using the common dtype from `start`, `end` and `periods` will give us an incorrect result. I realise that looping over all the bin_edges could be expensive so I'm pushing the code for now but trying out some alternatives.

Reply:
What is the output dtype of the `sequence` call? I imagine it isn't always `float`?

Reply:
I believe it's a column that takes on the dtype of the scalars that you use as inputs. In our case I'm using 
`start = cudf.Scalar(start, dtype="float64").device_value` and `freq_step = cudf.Scalar(freq_step, dtype="float64").device_value` . 

Using `int` as the scalar dtype can distort the result as sometimes you might put in two integers but their bin_edges may be floats.

Reply:
As an example using my current code and calling `interval_range`, I get the following from `sequence`

```
cudf.interval_range(start=0,end=10, periods=2,closed='left')

-> bin_edges = sequence(size=periods + 1, init=start, step=freq_step,)
(Pdb) bin_edges
<cudf.core.column.numerical.NumericalColumn object at 0x7f93e22c55f0>
[
  0,
  5,
  10
]
dtype: float64
(Pdb) 
````

Reply:
Something is up here, because libcudf types the output of `sequence` based on the type of `init`: https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/src/filling/sequence.cu#L67

Reply:
Let's continue discussion here: https://github.com/rapidsai/cudf/pull/7182#discussion_r604998435